catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/data/impl/MetacardImpl.java

```diff
@@ -715,11 +716,25 @@ public void setAttribute(String name, Serializable value) {
 
     @Override
     public void setAttribute(Attribute attribute) {
-
         if (attribute == null) {
             return;
         }
 
+        AttributeDescriptor descriptor = this.type.getAttributeDescriptor(attribute.getName());
+        if (descriptor != null) {
+            Class binding = descriptor.getType().getBinding();
+            if (binding != null) {
+                AttributeType.AttributeFormat format = descriptor.getType().getAttributeFormat();
+                if (attribute.getValue() != null && !attribute.getValue().getClass().equals(binding)) {
```

Comment:
From what I've seen, when it's dealing with primitives `isAssignableFrom` and `equals` are pretty close. The `binding.cast` below also adds a bit of permissiveness before throwing any exceptions.